tensorflow/compiler/xla/client/xla_builder.h

```diff
@@ -1038,6 +1056,37 @@ class XlaScopedShardingAssignment {
   absl::optional<OpSharding> prev_sharding_;
 };
 
+// RAII-style object: sets the current frontend attributes in builder on
+// construction, and clears it on destruction.
+class XlaScopedFrontendAttributesAssignment {
+ public:
+  XlaScopedFrontendAttributesAssignment(
+      xla::XlaBuilder* builder, absl::optional<FrontendAttributes> attributes)
+      : builder_(builder) {
+    SetFrontendAttributes(attributes);
+  }
+
+  XlaScopedFrontendAttributesAssignment(
+      const XlaScopedFrontendAttributesAssignment&) = delete;
+  XlaScopedFrontendAttributesAssignment& operator=(
+      const XlaScopedFrontendAttributesAssignment&) = delete;
+
+  ~XlaScopedFrontendAttributesAssignment() {
+    SetFrontendAttributes(absl::nullopt);
```

Comment:
Shouldn't this restore the frontend attributes to the value before it was set (i.e. before `XlaScopedFrontendAttributesAssignment::XlaScopedFrontendAttributesAssignment`)?

Reply:
That's a good point, I guess we need to decide how these attributes are getting set.
Currently, we use them internally to set attributes at the python scope level that will affect the operations added in that scope, so we do something like:
```python
@tf_contextlib.contextmanager
def frontend_attribute(attribute_name, attribute_value):

  proto = xla_data_pb2.FrontendAttributes()
  # pylint: disable=protected-access
  attr_value = ops.get_default_graph()._attr_scope_map.get(
      FRONTEND_ATTRIBUTES_NAME)
  # pylint: enable=protected-access
  new_attribute = True
  if attr_value:
    proto.ParseFromString(attr_value.s)
  proto.map[attribute_name] = attribute_value

  attr_value = attr_value_pb2.AttrValue(s=proto.SerializeToString())
  attrs = {FRONTEND_ATTRIBUTES_NAME: attr_value}

  # pylint: disable=protected-access
  with ops.get_default_graph()._attr_scope(attrs):
    yield
  # pylint: enable=protected-access
```
So basically the backup / restore of the previous attributes is done at the python level and in `XlaCompilationDevice::Compute` we could just do `builder->SetFrontendAttributes()` the ScopeAssignment is actually overkill.
Our C++ backend also adds some attributes, but this all happens further down the pipeline so it's not a problem here.

Now, I guess at some point we might want to set `FrontendAttributes` from the C++ runtime before `XlaCompilationDevice::Compute`.
If we consider this a valid use case then we would indeed need a ScopeAssignment but I think in that case it should probably do:

ScopeAssignment constructor:
- save existing cpp attributes.
- merge the python attributes with the cpp attributes and set them as the current attributes in the builder.

ScopeAssignment destructor:
- restore cpp attributes.

Otherwise with the current set/clear calls we will randomly erase the cpp attributes whenever some are set in python.

Reply:
> If we consider this a valid use case then we would indeed need a ScopeAssignment but I think in that case it should probably do:
> 
> ScopeAssignment constructor:
> 
> * save existing cpp attributes.
> * merge the python attributes with the cpp attributes and set them as the current attributes in the builder.

I don't think the merging behavior should be implemented by `XlaScopedFrontendAttributesAssignment`.  It should be implemented in its caller.

And `XlaScopedFrontendAttributesAssignment` should just save/restore frontend attributes without any intelligent merging behavior.

IOW the caller would do:

```
FrontendAttrs fa = GetCurrentAttrs();
FrontendAttrs new_fa = Merge(fa, attrs_from_node_def);
XlaScopedFrontendAttributesAssignment raii(std::move(new_fa));
```

Naively this would incur an extra copy, but that can be avoided by making `XlaScopedFrontendAttributesAssignment` move out the existing `FrontendAttrs` instead of copying it out.

> 
> ScopeAssignment destructor:
> 
> * restore cpp attributes.
> 
> Otherwise with the current set/clear calls we will randomly erase the cpp attributes whenever some are set in python.

Reply:
So basically you want to replace `XlaBuilder::SetFrontendAttributes` with a `XlaBuilder::SwapFrontendAttributes` that will return the old value in order to avoid the copy ?

Reply:
Yeah, that would work as well (and sounds simpler than my idea of adding a `FrontendAttributes&& ConsumeFrontendAttributes()`).

Reply:
Done